Refactor CPU embedding_bag implementation

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #18957 EmbeddingBag w/ differentiable per_sample_weights
* #18800 EmbeddingBag w/ per_sample_weights CUDA fwd + bwd
* #18799 EmbeddingBag w/ per_sample_weights CPU backward
* #18735 EmbeddingBag CPU forward with per_sample_weights.
* **#18734 Refactor CPU embedding_bag implementation**

In preparation for adding weighting.

Test Plan:
- Run tests

gh-metadata: pytorch pytorch 18734 gh/zou3519/25/head

Differential Revision: [D14851413](https://our.internmc.facebook.com/intern/diff/D14851413)

zou3519 committed:
Refactor CPU embedding_bag implementation

In preparation for adding weighting.

Test Plan:
- Run tests

zou3519 commented:
@pytorchbot retest this please

zou3519 committed:
Update on "Refactor CPU embedding_bag implementation"

Refactor CPU embedding_bag implementation

In preparation for adding weighting.

Test Plan:
- Run tests

gh-metadata: pytorch pytorch 18734 gh/zou3519/25/head

cpuhrsch commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -237,6 +235,62 @@ Tensor _embedding_bag_backward(const Tensor &grad, const Tensor &indices,
   }
 }
 
+static Tensor _embedding_bag_dense_backward_cpu_max(
```
I'm assuming this is just a copy-paste of a block of code previously part of _embedding_bag_dense_backward_cpu.

zou3519 replied:
yes

cpuhrsch commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -237,6 +235,62 @@ Tensor _embedding_bag_backward(const Tensor &grad, const Tensor &indices,
+  return index_grad_weight;
+}
+
+static std::vector<int64_t> compute_counts(
```
I'm assuming this is just a copy-paste of a block of code previously part of _embedding_bag_dense_backward_cpu.

zou3519 replied:
yes

cpuhrsch commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -264,74 +327,41 @@ Tensor _embedding_bag_dense_backward_cpu(const Tensor &grad_, const Tensor &indi
+  auto next_unique_index_idx =
+      compute_counts_uniq(num_weights, indices_data, numel, counts);
+
+  #pragma omp parallel for if (numel > 1000)
```
It's preferable to use parallel.h, if you want to fix more issues with this code.

zou3519 replied:
I don't know what parallel.h is, but this was here before and this pull request is a refactor of the code

cpuhrsch replied:
Yes, that was added after the last time I touched this file: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/Parallel.h

zou3519 replied:
Going to bootcamp this...

cpuhrsch commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -264,74 +327,41 @@ Tensor _embedding_bag_dense_backward_cpu(const Tensor &grad_, const Tensor &indi
+        AT_DISPATCH_FLOATING_TYPES(grad.scalar_type(), "embedding_bag_backward", [&] {
+          auto igwd = index_grad_weight.data<scalar_t>();
+          auto gd = grad.data<scalar_t>();
+          THBlas_axpy<scalar_t>(ddim, (scalar_t)scale, gd + ddim * source, 1,
```
I wonder if this is still necessary. It was added to deal with overhead introduced by the corresponding ATen function. Generally, we, of course, want to get rid of legacy TH code. This could be another nix thing to fix.

zou3519 replied:
The reason why EmbeddingBag uses axpy, to my understanding, is that it was faster than the alternative. It might be that it doesn't work as well now, but we can figure that out in a future change.

zou3519 committed:
Update on "Refactor CPU embedding_bag implementation"

Refactor CPU embedding_bag implementation

In preparation for adding weighting.

Test Plan:
- Run tests

gh-metadata: pytorch pytorch 18734 gh/zou3519/25/head
